### PR TITLE
Update API and add support for small molecules

### PR DESCRIPTION
**Pull request by wfondrie**
This PR makes some pretty large API changes, but provide a lot more flexibility than before.

### Changed
- `PeptideTransformer*` is now `AnalyteTransformer*` to reflect support for small molecules as well.
- Similarly, `PeptideDataset` was renamed `AnalyteDataset`.
- All the `forward()` methods for the Transformer modules have been updated to with additional keyword arguments. This enables BERT-style masking for training and such.
- All transformer modules now have a `global_token_hook()` method. This can be overwritten in a subclass to customize how a global token (the first element of the sequence) is created using the `*args` and `**kwargs` provided in the forward methods.

### Added
- A new tokenizer for small molecules, `MoleculeTokenizer`.
- The `PeptideIonTokenizer.calculate_precursor_ions()` method is fully PyTorch, so it should be efficient for use during model training and inference. (see https://github.com/Noble-Lab/casanovo/pull/269)

Next PR will be docs!

### CODE REVIEW

File: depthcharge/transformers/analytes.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+import torch
+
+from .. import utils
+from ..encoders import PositionalEncoder
+from ..mixins import ModelMixin, TransformerMixin
+from ..tokenizers import MoleculeTokenizer, PeptideTokenizer
+
+
+class _AnalyteTransformer(torch.nn.Module, ModelMixin, TransformerMixin):
+    """A transformer base class for peptide sequences.
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:11]: **nitpick:** peptide -> analyte

File: depthcharge/transformers/analytes.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+    Parameters
+    ----------
+    n_tokens : int or PeptideTokenizer
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:15]: **nitpick:** `PeptideTokenizer` -> `Tokenizer` base class.

File: depthcharge/transformers/analytes.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        The number of tokens used to tokenize peptide sequences.
+    d_model : int
+        The latent dimensionality to represent the amino acids in a peptide
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:18]: **nitpick:** Generalize to "tokens in a molecular sequence".

File: depthcharge/transformers/analytes.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        sequence.
+    positional_encoder : PositionalEncoder or bool
+        The positional encodings to use for the amino acid sequence. If
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:21]: **nitpick:** "amino acid sequence" -> "molecular sequence"

File: depthcharge/transformers/analytes.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        ``True``, the default positional encoder is used. ``False`` disables
+        positional encodings, typically only for ablation tests.
+    max_charge : int
+        The maximum precursor charge to embed.
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:25]: **question:** Now that we're venturing into small molecule land, do we support negative charges as well? I assume we currently also default to protonated charges only?

File: depthcharge/transformers/analytes.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+    """
+
+    def __init__(
+        self,
+        n_tokens: int | PeptideTokenizer | MoleculeTokenizer,
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:30]: **suggestion:** Generalize to `Tokenizer`.

File: depthcharge/transformers/analytes.py (hunk 5 above)
Comment (bittremieux) [reply depthcharge/transformers/analytes.py:25]: **issue:** This doesn't actually seem to be a parameter in the code.

File: depthcharge/transformers/analytes.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        d_model: int,
+        nhead: int,
+        dim_feedforward: int,
+        n_layers: int,
+        dropout: float,
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:35]: **issue:** Several missing docstrings.

File: depthcharge/transformers/analytes.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+            self.positional_encoder = positional_encoder
+        elif positional_encoder:
+            self.positional_encoder = PositionalEncoder(d_model)
+        else:
+            self.positional_encoder = torch.nn.Identity()
+
+        self.token_encoder = torch.nn.Embedding(
+            n_tokens + 1,
+            d_model,
+            padding_idx=0,
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:61]: **question:** Should this now use the `padding_int` that's included in the tokenizer classes?

File: depthcharge/transformers/analytes.py [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+        )
+
+    def global_token_hook(
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:64]: **praise:** This could be quite powerful. Nice. 🙂

File: depthcharge/transformers/analytes.py [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+        """Define how additional information in the batch may be used.
+
+        Overwrite this method to define custom functionality dependent on
+        information in the batch. Examples would be to incorporate any
+        combination of the mass, charge, retention time, or
+        ion mobility of a peptide.
+
+        The representation returned by this method is preprended to the
+        peak representations that are fed into the Transformer and
+        ultimately contribute to the peptide representation that is the
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:79]: **nitpick:** "peptide" -> analyte

File: depthcharge/transformers/analytes.py [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+        first element of the sequence in the model output.
+
+        By default, this method returns a tensor of zeros.
+
+        Parameters
+        ----------
+        tokens : list of str, torch.Tensor, or None
+            The partial peptide sequences for which to predict the next
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:87]: **nitpick:** "peptide" -> "molecular", next line "amino acid" -> "token".

File: depthcharge/transformers/analytes.py [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+            of a string.
+        *args : torch.Tensor
+            Additional data passed with the batch.
+        **kwargs : dict
+            Additional data passed with the batch.
+
+        Returns
+        -------
+        torch.Tensor of shape (batch_size, d_model)
+            The precursor representations.
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:98]: **typo:** Precursor is a typo I assume?

File: depthcharge/transformers/analytes.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+        tokens: torch.Tensor,
+        *args: torch.Tensor,
+        mask: torch.Tensor = None,
+        **kwargs: dict,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Encode a collection of sequences.
+
+        Parameters
+        ----------
+        tokens : torch.Tensor of size (batch_size, peptide_length)
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:177]: **nitpick:** peptide_length -> analyte_length

File: depthcharge/transformers/analytes.py [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+        information in the batch. Examples would be to incorporate any
+        combination of the mass, charge, retention time, or
+        ion mobility of a peptide.
+
+        The representation returned by this method is preprended to the
+        peak representations that are fed into the Transformer and
+        ultimately contribute to the peptide representation that is the
+        first element of the sequence in the model output.
+
+        By default, this method returns a tensor of zeros.
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:82]: **question:** By concatenating the zeros, does this change the default operation of the model compared to previously?

File: depthcharge/transformers/analytes.py [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+        return latent, src_key_padding_mask
+
+
+class AnalyteTransformerDecoder(_AnalyteTransformer):
+    """A transformer decoder for peptide or small molecule sequences.
+
+    Parameters
+    ----------
+    n_tokens : int, PeptideTokenizer, or MoleculeTokenizer
+        The number of tokens used to tokenize peptide sequences.
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:225]: **typo:** peptide -> molecular

File: depthcharge/transformers/analytes.py [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+        divisible by ``nhead``.
+    dim_feedforward : int, optional
+        The dimensionality of the fully connected layers in the Transformer
+        layers of the model.
+    n_layers : int, optional
+        The number of Transformer layers.
+    dropout : float, optional
+        The dropout probability for all layers.
+    positional_encoder : PositionalEncoder or bool, optional
+        The positional encodings to use for the amino acid sequence. If
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:239]: **typo:** amino acid -> molecular

File: depthcharge/transformers/analytes.py [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+        memory_mask: torch.Tensor | None = None,
+        tgt_mask: torch.Tensor | None = None,
+        **kwargs: dict,
+    ) -> torch.Tensor:
+        """Embed a collection of sequences.
+
+        Parameters
+        ----------
+        tokens : list of str, torch.Tensor, or None
+            The partial peptide sequences for which to predict the next
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:299]: **typo:** peptide -> molecular. Next line: amino acid -> token

File: depthcharge/transformers/analytes.py [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+        memory_mask: torch.Tensor | None = None,
+        tgt_mask: torch.Tensor | None = None,
+        **kwargs: dict,
+    ) -> torch.Tensor:
+        """Decode a collection of sequences.
+
+        Parameters
+        ----------
+        tokens : list of str, torch.Tensor, or None
+            The partial peptide sequences for which to predict the next
```
Comment (bittremieux) [inline depthcharge/transformers/analytes.py:394]: **typo:** peptide -> molecular. Next line: amino acid -> token

File: depthcharge/tokenizers/peptides.py [hunk 19]
```diff
@@ -4,110 +4,138 @@
...
-    Parse and tokenize ProForma-compliant peptide sequences. Additionally,
-    use this class to calculate fragment and precursor ion masses.
+    Parse and tokenize ProForma-compliant peptide sequences.
 
     Parameters
     ----------
-    residues : dict[str, float], optional
-        Residues and modifications to add to the vocabulary beyond the
-        standard 20 amino acids.
+    residues : iterable of str
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply depthcharge/tokenizers/peptides.py:23]: Good catch!

File: depthcharge/transformers/analytes.py (hunk 5 above)
Comment (wfondrie) [reply depthcharge/transformers/analytes.py:25]: Ah yes! This now must be handled by the user with the `global_token_hook`.

File: depthcharge/transformers/analytes.py (hunk 14 above)
Comment (wfondrie) [reply depthcharge/transformers/analytes.py:82]: I guess it depends on what you mean by previously 🤔 - however, maybe it would be better not to concatenate anything?

Comment (wfondrie) [reply depthcharge/transformers/analytes.py:82]: Actually, I define it padding in the `tgt_key_padding_mask`, so it is ignored already.

File: depthcharge/tokenizers/molecules.py [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            The SMILES or SELFIES string representing a molecule.
+
+        Returns
+        -------
+        List[str]
+            The SELFIES tokens representing the molecule.
+        """
+        try:
+            return sf.split_selfies(sequence)
+        except AttributeError:
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply depthcharge/tokenizers/molecules.py]: Good catch here! There definitely was a bug 😄

File: depthcharge/tokenizers/molecules.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+        stop_token : str, optional
+            The stop token to use.
+
+        Returns
+        -------
+        MoleculeTokenizer
+            The tokenizer restricted to the vocabulary present in the
+            input SMILES strings.
+        """
+        vocab = sf.get_alphabet_from_selfies(
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply depthcharge/tokenizers/molecules.py:102]: It'll yield a `selfies.EncoderError`.

**Comment by wfondrie** [comment]
Thanks a lot @bittremieux! It turns out a missed a lot of documentation and you caught at least two potential bugs 🎉 

Sorry the update has so many new changes - Ruff updated and apparently added another rule to the autoformatting.

**Comment by wfondrie** [comment]
I'm going to go ahead and merge so I can build atop this. @bittremieux, if there's anything else please let me know and I'll address it in a fresh PR!
